pkg/authorization/authz.go

```diff
@@ -0,0 +1,130 @@
+package authorization
+
+import (
+	"fmt"
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/docker/pkg/plugins"
+	"github.com/docker/docker/pkg/version"
+	"net/http"
+	"net/http/httptest"
+	"net/http/httputil"
+)
+
+// NewCtx creates new authZ context, it is used to store authorization information related to a specific docker
+// REST http session
+// A context provides two method:
+// Authenticate Request:
+// Call authZ plugins with current REST request and AuthN response
+// Request contains full HTTP packet sent to the docker daemon
+// https://docs.docker.com/reference/api/docker_remote_api/
+//
+// Authenticate Response:
+// Call authZ plugins with full info about current REST request, REST response and AuthN response
+// The response from this method may contains content that overrides the daemon response
+// This allows authZ plugins to filter privileged content
+//
+// If multiple authZ plugins are specified, the block/allow decision is based on ANDing all plugin results
+// For response manipulation, the response from each plugin is piped between plugins. Plugin execution order
+// is determined according to daemon parameters
+func NewCtx(authZPlugins []string, user, userAuthNMethod, requestMethod, requestUri string) *Ctx {
+	return &Ctx{authZPlugins: authZPlugins, user: user, userAuthNMethod: userAuthNMethod, requestMethod: requestMethod, requestUri: requestUri}
+}
+
+type Ctx struct {
+	user            string
+	userAuthNMethod string
+	requestMethod   string
+	requestUri      string
+	authZPlugins    []string
+	plugins         []*plugins.Plugin
+	rawRequest      []byte
+}
+
+// AuthRequest authorized the request to the docker daemon using authZ plugins
+func (a *Ctx) AuthZRequest(w http.ResponseWriter, r *http.Request) (err error) {
+
+	logrus.Debugf("AuthZ requset using plugins %s", a.authZPlugins)
+
+	for _, name := range a.authZPlugins {
+
+		p, err := plugins.Get(name, AuthZApiImplements)
+		if err != nil {
+			return err
+		}
+		a.plugins = append(a.plugins, p)
```

Comment:
Is it necessary to parse and build this on every request? Could instead the `Ctx` struct be built with a set of `authZPlugins` of type `[]*plugins.Plugin`? Or better yet does it make sense to create a typed plugin that could be plugins.AuthZPlugin`?